network/crl.go

```diff
@@ -120,19 +149,124 @@ type CRLVerifier interface {
+		log.Debugln("CRL: Skipping check since no config URI specified and we were told to ignore URIs from certificate")
 		return 0, nil
 	}
 
-	log.Infof("CRL: Verify( %s )", chain[0].Subject.CommonName)
+	cert := chain[0]
+	issuer := chain[1]
+
+	for i := range cert.CRLDistributionPoints {
+		log.Debugf("CRL: certificate contains CRL URI: %s", cert.CRLDistributionPoints[i])
+	}
+
+	confirmsByConfigCRL := 0
+	confirmsByCertCRL := 0
+
+	// TODO avoid querying same CRL more than once, maybe create some list of checked CRLs (based on URI)
+
+	for {
+		if len(v.Config.uri) > 0 {
+			rawCRL, err := v.getCachedOrFetch(v.Config.uri)
+			if err != nil {
+				log.WithError(err).Debugf("CRL: Cannot get CRL from '%s'", v.Config.uri)
+				// XXX return error instead?
+				break
+			}
+
+			crl, err := x509.ParseCRL(rawCRL)
+			if err != nil {
+				log.WithError(err).Debugf("CRL: Cannot parse CRL from '%s'", v.Config.uri)
+				// XXX return error instead?
+				break
+			}
+
+			err = issuer.CheckCRLSignature(crl)
+			if err != nil {
+				log.WithError(err).Warnf("CRL: Failed to check signature for CRL at %s", v.Config.uri)
+				// XXX return error instead?
+				break
+			}
+
+			for i := range crl.TBSCertList.RevokedCertificates {
+				if crl.TBSCertList.RevokedCertificates[i].SerialNumber.Cmp(cert.SerialNumber) == 0 {
+					log.Warnf("CRL: Certificate %v was revoked at %v", cert.SerialNumber, crl.TBSCertList.RevokedCertificates[i].RevocationTime)
+					return 0, errors.New(fmt.Sprintf("Certificate %v was revoked at %v", cert.SerialNumber, crl.TBSCertList.RevokedCertificates[i].RevocationTime))
+				}
+			}
+
+			confirmsByConfigCRL += 1
+		}
+		break
+	}
+
+	if len(cert.CRLDistributionPoints) > 0 {
+		for i := range cert.CRLDistributionPoints {
+			rawCRL, err := v.getCachedOrFetch(cert.CRLDistributionPoints[i])
+			if err != nil {
+				log.WithError(err).Debugf("CRL: Cannot get CRL from '%s'", cert.CRLDistributionPoints[i])
+				// XXX return error instead?
+				continue
+			}
+
+			crl, err := x509.ParseCRL(rawCRL)
+			if err != nil {
+				log.WithError(err).Debugf("CRL: Cannot parse CRL from '%s'", cert.CRLDistributionPoints[i])
+				// XXX return error instead?
+				continue
+			}
+
+			err = issuer.CheckCRLSignature(crl)
+			if err != nil {
+				log.WithError(err).Warnf("CRL: Failed to check signature for CRL at %s", cert.CRLDistributionPoints[i])
+				// XXX return error instead?
+				continue
+			}
+
+			if crl.TBSCertList.RevokedCertificates[i].SerialNumber.Cmp(cert.SerialNumber) == 0 {
+				log.Warnf("CRL: Certificate %v was revoked at %v", cert.SerialNumber, crl.TBSCertList.RevokedCertificates[i].RevocationTime)
+				return 0, errors.New(fmt.Sprintf("Certificate %v was revoked at %v", cert.SerialNumber, crl.TBSCertList.RevokedCertificates[i].RevocationTime))
+			}
+		}
+
+		confirmsByCertCRL += 1
+	}
+
+	// XXX with cache containing raw CRL, we have to parse and verify signature on every check,
```

Comment:
nice idea, and better to use some LRU cache, to store only frequently requested CRL values. And we already have one [implementation](https://github.com/cossacklabs/acra/blob/master/keystore/lru/cache.go#L29) for that, used in keystore.
And I looked on lru implementation [doc](https://github.com/golang/groupcache/blob/master/lru/lru.go#L25) and see that it allow having unlimited size 
So maybe we can use our lru cache instead new one DefaultCRLCache? What do you think?

Reply:
`CRLCache` will still have `[]byte` as value type, so maybe we need to change the interface? Or introduce one more interface to use between `CRLCache` and `CRLVerifier` and store some amount of parsed-verified CRLs

Reply:
Or... Did you mean add the LRU cache into `DefaultCRLVerifier` itself? Then I will only need to extend `DefaultCRLVerifier.getCachedOrFetch()` a little

Reply:
> Or introduce one more interface to use between CRLCache and CRLVerifier and store some amount of parsed-verified CRLs

hm... actually it okay to provide one more interface for CRL/OCSP, which will wrap cache implementation and return typed values, to avoid extra parsing

> Did you mean add the LRU cache into DefaultCRLVerifier itself? Then I will only need to extend DefaultCRLVerifier.getCachedOrFetch() a little

not only add into, but better to hide DefaultCRLVerifier struct, probably make it private (to avoid misuse and allow initialize only by constructors) and add [factory method](https://github.com/cossacklabs/acra/blob/master/keystore/lru/cache.go#L45) which will accept size of cache as parameter and will initialize all that it need.